### PR TITLE
Implement polls

### PR DESCRIPTION
**Pull request by panpawn**
This is a plugin that allows each room to have a poll system.  In doing such, this will allow for rooms to have another scripted method user interface that the user can interact with while also voicing their opinions with in the method of a "survey", so to speak.

### CODE REVIEW

**Comment by Rhythms88** [comment]
There used to be something like this iirc, but it was removed because people were switching alts and voting repeatedly.

**Comment by panpawn** [comment]
I think I eliminated that issue because it checks IP.

**Comment by Ridaz** [comment]
Why is this needed? People can just use straw poll.

**Comment by Rhythms88** [comment]
^ that too

**Comment by panpawn** [comment]
Why make people go to another website to vote if the feature can easily be implemented on PS! is a better question imo

**Comment by Ridaz** [comment]
Because it demands more of server

**Comment by panpawn** [comment]
It doesn't really do anything that uses many resources that I feel would make this pull request any less valid in my opinion.

**Comment by Rhythms88** [comment]
Also why is it allowing voices to host these

**Comment by panpawn** [comment]
That can be changed pretty easily, that's just what rank I felt would be a good option to begin with until further input was given.

**Comment by ScottehMax** [comment]
Did you try running `npm test` before committing...?

**Comment by panpawn** [comment]
No, I have not

**Comment by Rhythms88** [comment]
Always run `npm test` before submitting Pull Requests.

**Comment by Zarel** [comment]
@panpawn is right, PS should indeed have a built-in poll system.

I'd like to make my own with a GUI, though.

**Comment by kotarou3** [comment]
There needs to be client sided component to this to make them look nicer.
E.g., Some CSS.

File: chat-plugins/polls.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+	votes: function (target, room, user) {
+		if (!room.answers) room.answers = new Object();
+		if (!room.question) return this.sendReply('There is no poll currently going on in this room.');
+		if (!this.canBroadcast()) return;
+		this.sendReply('NUMBER OF VOTES: ' + Object.keys(room.answers).length);
+	},
+
+	vote: function (target, room, user) {
+		var ips = JSON.stringify(user.ips);
```
Comment (kotarou3) [inline chat-plugins/polls.js]: ?
This is unreliable

File: chat-plugins/polls.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+		for (var i in obj.answerList) options[obj.answerList[i]] = 0;
+		for (var i in obj.answers) options[obj.answers[i]]++;
+		var sortable = new Array();
+		for (var i in options) sortable.push([i, options[i]]);
+		sortable.sort(function(a, b) {return a[1] - b[1]});
+		var html = "";
+		for (var i = sortable.length - 1; i > -1; i--) {
+			var option = sortable[i][0];
+			var value = sortable[i][1];
+			if (value > 0) html += "&bull; " + Tools.escapeHTML(option) + " - " + Math.floor(value / votes * 100) + "% (" + value + ")<br />";
```
Comment (kotarou3) [inline chat-plugins/polls.js]: Use `<ul><li>...</li><li>...</li>...</ul>`

File: chat-plugins/polls.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+			room.answers = new Object();
+			return room.addRaw("<h3>The poll was canceled because of lack of voters.</h3>");
+		}
+		var options = new Object();
+		var obj = Rooms.get(room);
+		for (var i in obj.answerList) options[obj.answerList[i]] = 0;
+		for (var i in obj.answers) options[obj.answers[i]]++;
+		var sortable = new Array();
+		for (var i in options) sortable.push([i, options[i]]);
+		sortable.sort(function(a, b) {return a[1] - b[1]});
```
Comment (kotarou3) [inline chat-plugins/polls.js]: Why not just sort `Object.keys(options)`?

File: chat-plugins/polls.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+			output += '<button name="send" value="/vote '+room.answerList[u]+'">'+Tools.escapeHTML(room.answerList[u])+'</button>&nbsp;';
+		}
+		this.sendReply('|raw|<div class="infobox"><h2>' + Tools.escapeHTML(room.question) + separacion + '<font font size=1 color = "#939393"><small>/vote OPTION</small></font></h2><hr />' + separacion + separacion + output + '</div>');
+	},
+
+	ep: 'endpoll',
+	endpoll: function (target, room, user) {
+		if (!this.canBroadcast()) return;
+		if ((user.locked || user.mutedRooms[room.id]) && !user.can('bypassall')) return this.sendReply("You cannot do this while unable to talk.");
+		if (!room.question) return this.sendReply('There is no poll to end in this room.');
```
Comment (kotarou3) [inline chat-plugins/polls.js]: Use double quotes here

File: chat-plugins/polls.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	survey: 'poll',
+	poll: function (target, room, user) {
+		if (!this.canBroadcast()) return;
+		if ((user.locked || user.mutedRooms[room.id]) && !user.can('bypassall')) return this.sendReply("You cannot do this while unable to talk.");
+		if (room.question) return this.sendReply('There is currently a poll going on already.');
+		if (!target) return false;
+		if (target.length > 500) return this.sendReply('Polls can not be this long.');
+		var separacion = "&nbsp;&nbsp;";
+		var answers = splint(target);
+		if (answers.length < 3) return this.sendReply('Correct syntax for this command is /poll question, option, option...');
```
Comment (kotarou3) [inline chat-plugins/polls.js]: Put this in `/pollhelp` and call `/help poll` from here instead

File: chat-plugins/polls.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+exports.commands = {
+
+	survey: 'poll',
+	poll: function (target, room, user) {
+		if (!this.canBroadcast()) return;
+		if ((user.locked || user.mutedRooms[room.id]) && !user.can('bypassall')) return this.sendReply("You cannot do this while unable to talk.");
+		if (room.question) return this.sendReply('There is currently a poll going on already.');
+		if (!target) return false;
+		if (target.length > 500) return this.sendReply('Polls can not be this long.');
```
Comment (kotarou3) [inline chat-plugins/polls.js]: This limit is way too high

File: chat-plugins/polls.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+	ep: 'endpoll',
+	endpoll: function (target, room, user) {
+		if (!this.canBroadcast()) return;
+		if ((user.locked || user.mutedRooms[room.id]) && !user.can('bypassall')) return this.sendReply("You cannot do this while unable to talk.");
+		if (!room.question) return this.sendReply('There is no poll to end in this room.');
+		if (!room.answers) room.answers = new Object();
+		var votes = Object.keys(room.answers).length;
+		if (votes == 0) {
+			room.question = undefined;
```
Comment (kotarou3) [inline chat-plugins/polls.js]: Set it to either `null` or `""`, never `undefined`

File: chat-plugins/polls.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+			if (value > 0) html += "&bull; " + Tools.escapeHTML(option) + " - " + Math.floor(value / votes * 100) + "% (" + value + ")<br />";
+		}
+		room.addRaw('<div class="infobox"><h2>Results to "' + Tools.escapeHTML(obj.question) + '"<br /><i><font size=1 color = "#939393">Poll ended by ' + Tools.escapeHTML(user.name) + '</font></i></h2><hr />' + html + '</div>'); 
+		room.question = undefined;
+		room.answerList = new Array();
+		room.answers = new Object();
+	},
+
+	votes: function (target, room, user) {
+		if (!room.answers) room.answers = new Object();
```
Comment (kotarou3) [inline chat-plugins/polls.js]: ?

File: chat-plugins/polls.js [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+		var separacion = "&nbsp;&nbsp;";
+		var answers = splint(target);
+		if (answers.length < 3) return this.sendReply('Correct syntax for this command is /poll question, option, option...');
+		var question = answers[0];
+		question = Tools.escapeHTML(question);
+		answers.splice(0, 1);
+		var answers = answers.join(',').toLowerCase().split(',');
+		room.question = question;
+		room.answerList = answers;
+		room.usergroup = Config.groupsranking.indexOf(user.group);
```
Comment (kotarou3) [inline chat-plugins/polls.js]: Why not just `Config.groups[user.group].rank`?

File: chat-plugins/polls.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+exports.commands = {
+
+	survey: 'poll',
+	poll: function (target, room, user) {
+		if (!this.canBroadcast()) return;
+		if ((user.locked || user.mutedRooms[room.id]) && !user.can('bypassall')) return this.sendReply("You cannot do this while unable to talk.");
+		if (room.question) return this.sendReply('There is currently a poll going on already.');
+		if (!target) return false;
```
Comment (kotarou3) [inline chat-plugins/polls.js]: Silent failure? Bad. Call `/help poll` instead

File: chat-plugins/polls.js [hunk 11]
```diff
@@ -0,0 +1,102 @@
+/* Poll commands plugin
+ * This is a poll commands system that will allow a voice (or roomvoice) to host 
+ * polls in terms of starting, reminding
+ * by: panpawn
+ */
+
+function splint(target) {
```
Comment (kotarou3) [inline chat-plugins/polls.js]: You only ever use this function in one place, and it's short, so you should inline it

File: chat-plugins/polls.js [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+	var cmdArr = target.split(",");
+	for (var i = 0; i < cmdArr.length; i++) cmdArr[i] = cmdArr[i].trim();
+		return cmdArr;
+}
+
+exports.commands = {
+
+	survey: 'poll',
+	poll: function (target, room, user) {
+		if (!this.canBroadcast()) return;
```
Comment (kotarou3) [inline chat-plugins/polls.js]: Do a `this.can('poll', null, room)` instead

File: chat-plugins/polls.js [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+		if ((user.locked || user.mutedRooms[room.id]) && !user.can('bypassall')) return this.sendReply("You cannot do this while unable to talk.");
+		if (room.question) return this.sendReply('There is currently a poll going on already.');
+		if (!target) return false;
+		if (target.length > 500) return this.sendReply('Polls can not be this long.');
+		var separacion = "&nbsp;&nbsp;";
```
Comment (kotarou3) [inline chat-plugins/polls.js]: Spelling

File: chat-plugins/polls.js [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+		var answers = answers.join(',').toLowerCase().split(',');
+		room.question = question;
+		room.answerList = answers;
+		room.usergroup = Config.groupsranking.indexOf(user.group);
+		var output = '';
+		for (var u in room.answerList) {
+			if (!room.answerList[u] || room.answerList[u].length < 1) continue;
+			output += '<button name="send" value="/vote ' + room.answerList[u] + '">' + Tools.escapeHTML(room.answerList[u]) + '</button>&nbsp;';
+		}
+		room.addRaw('<div class="infobox"><h2>' + Tools.escapeHTML(room.question) + separacion + '<font size=2 color = "#939393"><small>/vote OPTION<br /><i><font size=1>Poll started by ' + Tools.escapeHTML(user.name) + '</font size></i></small></font></h2><hr />' + separacion + separacion + output + '</div>');
```
Comment (kotarou3) [inline chat-plugins/polls.js]: For most of the uses I've seen where `<br />` have been used on showdown, it's the wrong use for it.
Instead, use block elements instead. e.g., `<p>`, `<h1>`, `<div>`

File: chat-plugins/polls.js [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+	},
+
+	pr: 'pollremind',
+	pollremind: function (target, room, user) {
+		var separacion = "&nbsp;&nbsp;";
```
Comment (kotarou3) [inline chat-plugins/polls.js]: Spelling

**Comment by kotarou3** [comment]
Consider putting everything under the `poll` namespace, and not polluting the `Room` object

**Comment by Irraquated** [comment]
Actually @panpawn I was talking with Joim the other day about adding this myself. Looks nice.

File: chat-plugins/polls.js (hunk 6 above)
Comment (Freigeist) [inline chat-plugins/polls.js]: Also should be a `const` at begin of file

**Comment by panpawn** [comment]
I'm leaving this open as a reminder for Zarel to implement polls when he gets the chance.

**Comment by AustinXII** [comment]
@panpawn this should probably be closed ...

**Comment by Slayer95** [comment]
No. This is just an alternate implementation of polls, and not necessarily inferior as far as I can see.

Refs #2177

**Comment by Zarel** [comment]
New one takes advantage of the updatable HTML box API, and is being actively maintained.

Whether or not the new one ends up being merged, this one definitely won't, so I'm closing.

**Pull request closed without merging**